File: Smallest-Good-Base/SGB.py
import math

test1 = "13"
test2 = "4681"
test3 = "1000000000000000000"


class Solution(object):

    def smallestGoodBase(self, n):
        """
        :type n: str
        :rtype: str
        """
        n = int(n)
        for l in range(int(math.log(n, 2)) + 1, 2, -1):
            k = int(math.pow(n, 1 / (l - 1)))
            print(l, k)
            if (k ** l - 1) / (k - 1) == n:
                return str(k)
        return str(n - 1)

File: Smallest-Good-Base/SGB.hs
-- well, I give up
-- RealFrac is a weird typeclass, or becasue I am still suck with haskell
log' :: (Integral a, RealFrac a, Floating a) => a -> a
log' = floor.(+1).logBase 2

pow' :: (Integral b, RealFrac a, Integral a) => a -> a -> b
pow' n l = floor (n ^ (1 / (l-1)))

checkAnswer :: (Integral b) => b -> b -> b
checkAnswer k l = (k ^ l - 1) `div` (k - 1)


{--
smallGoodBase :: (RealFrac a, Integral a) => a -> a -> a
smallGoodBase n l
  | l <= 2
  = n - 1
  | otherwise
  = let k = (pow' n l) :: Int in
      if checkAnswer k l == n then k
      else smallGoodBase n (l - 1)
--}

{--
main = do
  let test1 = 13
      test2 = 4681
      test3 = 1000000000000000000
  print (smallGoodBase test1)
--}
